streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractTransactionalStore.java

```diff
@@ -0,0 +1,442 @@
+                                    segmentId,
+                                    metricsRecorder);
+    }
+
+    public abstract T mainStore();
+
+    public abstract KeyValueSegment tmpStore();
+
+    @Deprecated
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        if (context instanceof StateStoreContext) {
+            init((StateStoreContext) context, root);
+        } else {
+            throw new UnsupportedOperationException(
+                "Use TransactionalKeyValueStore#init(StateStoreContext, StateStore) instead."
+            );
+        }
+    }
+
+    @Override
+    public void init(final StateStoreContext context, final StateStore root) {
+        this.context = context;
+
+        doInit(context.appConfigs(), context.stateDir());
+        ((RocksDBStore) mainStore()).openDB(configs, stateDir);
+
+        final File positionCheckpointFile = new File(context.stateDir(), name() + ".position");
+        this.positionCheckpoint = new OffsetCheckpoint(positionCheckpointFile);
+        this.position = StoreQueryUtils.readPositionFromCheckpoint(positionCheckpoint);
+        tmpStore().consistencyEnabled = consistencyEnabled;
+
+        // register and possibly restore the state from the logs
+        context.register(
+            root,
+            (RecordBatchingStateRestoreCallback) this::restoreBatch,
+            () -> StoreQueryUtils.checkpointPosition(positionCheckpoint, position)
+        );
+
+        consistencyEnabled = StreamsConfig.InternalConfig.getBoolean(
+            context.appConfigs(),
+            IQ_CONSISTENCY_OFFSET_VECTOR_ENABLED,
+            false);
+    }
+
+    private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> records) {
+        final Collection<ConsumerRecord<byte[], byte[]>> changelogRecords = records
+            .stream()
+            .map(record -> new ConsumerRecord<>(
+                record.topic(),
+                record.partition(),
+                record.offset(),
+                record.key(),
+                toUncommittedValue(record.value())))
+            .collect(Collectors.toList());
+        tmpStore().restoreBatch(changelogRecords);
+        commit(null);
+    }
+
+    void doInit(final Map<String, Object> configs, final File stateDir) {
+        this.configs = configs;
+        this.stateDir = stateDir;
+        tmpStore().openDB(configs, stateDir);
+    }
+
+    @Override
+    public synchronized void close() {
+        final HashSet<KeyValueIterator<Bytes, byte[]>> iterators;
+        synchronized (openIterators) {
+            iterators = new HashSet<>(openIterators);
+        }
+        for (final KeyValueIterator<Bytes, byte[]> iterator : iterators) {
+            iterator.close();
+        }
+
+        tmpStore().close();
+        mainStore().close();
+    }
+
+    @Override
+    public void commit(final Long changelogOffset) {
+        tmpStore().commit(changelogOffset);
+        doCommit();
+    }
```

Comment:
Added missing `mainStore#commit` in [81f248b](https://github.com/apache/kafka/pull/12393/commits/81f248b19870e9438969b5fe7f28780f1bf1c538).

I like your idea to decouple checkpointing interval from commit interval! How about we address it in a follow-up issue? This patch is getting rather big and I would like to limit its scope at this point.

Reply:
I like the idea of decoupling rocksdb flush from the kafka producer commit because a slow commit/flush process would greatly reduce the feasibility of having a small commit interval (i.e. 50ms) in latency-sensitive applications.

However, one interesting side effect of coupling the producer.commit() with tmpStore.commit() is that it would enable a later KIP which allows Interactive Queries to only read committed values by directing the IQ to only the main store. As we stand, IQ can read uncommitted values from the store (generally that's not a _huge_ problem since most processing is deterministic).

Reply:
> How about we address it in a follow-up issue?
Good idea. It introduces a new configuration option, which will require a KIP.